test/lint/lint-filenames-permissions.py

```diff
@@ -0,0 +1,160 @@
+CMD_SHEBANG_FILES = """git grep --full-name --line-number -I '^#!'"""
+ALLOWED_FILENAME_REGEXP = "^[a-zA-Z0-9/_.@][a-zA-Z0-9/_.@-]*$"
+ALLOWED_SOURCE_FILENAME_REGEXP = "^[a-z0-9_./-]+$"
+ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP = (
+    "^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)"
+)
+ALLOWED_PERMISSION_NON_EXECUTABLES = 644
+ALLOWED_PERMISSION_EXECUTABLES = 755
+ALLOWED_EXECUTABLE_SHEBANG = {
+    "py": [b"#!/usr/bin/env python3"],
+    "sh": [b"#!/usr/bin/env bash", b"#!/bin/sh"],
+}
+
+
+def check_all_filenames():
+    """
+    Checks every file in the repository against an allowed regexp to make sure only lowercase or uppercase
+    alphanumerics (a-zA-Z0-9), underscores (_), hyphens (-), at (@) and dots (.) are used in repository filenames.
+    """
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").strip().split("\n")
+    filename_regex = re.compile(ALLOWED_FILENAME_REGEXP)
+    exit_code = 0
+    for filename in filenames:
+        if not filename_regex.match(filename):
+            print(
+                f"""File "{filename}" does not not match the allowed filename regexp ('{ALLOWED_FILENAME_REGEXP}')."""
+            )
+            exit_code = 1
+    return exit_code
+
+
+def check_source_filenames():
+    """
+    Checks only source files (*.cpp, *.h, *.py, *.sh) against a stricter allowed regexp to make sure only lowercase
+    alphanumerics (a-z0-9), underscores (_), hyphens (-) and dots (.) are used in source code filenames.
+
+    Additionally there is an exception regexp for directories or files which are excepted from matching this regexp.
+    """
+    filenames = check_output(CMD_SOURCE_FILES, shell=True).decode("utf8").strip().split("\n")
+    filename_regex = re.compile(ALLOWED_SOURCE_FILENAME_REGEXP)
+    filename_exception_regex = re.compile(ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP)
+    exit_code = 0
+    for filename in filenames:
+        if not filename_regex.match(filename) and not filename_exception_regex.match(filename):
+            print(
+                f"""File "{filename}" does not not match the allowed source filename regexp ('{ALLOWED_SOURCE_FILENAME_REGEXP}'), or the exception regexp ({ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP})."""
+            )
+            exit_code = 1
+    return exit_code
+
+
+def check_all_file_permissions():
+    """
+    Checks all files in the repository match an allowed executable or non-executable file permission octal.
+
+    Additionally checks that for executable files, the file contains a shebang line
+    """
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").strip().split("\n")
+    exit_code = 0
+    for filename in filenames:
+        file_extension = os.path.splitext(filename)[1].strip(".")
+        file_perms = int(oct(os.stat(filename).st_mode)[-3:])
+
+        if file_perms == ALLOWED_PERMISSION_EXECUTABLES:
+            shebang = open(filename, "rb").readline().strip()
```

Comment:
I believe the `strip()` here is not correct. I don't think that you can put whitespace before shebang. 

It's a question whether one should allow whitespace after shebank AFAIK it's allowed but ... It may be for the best simply remove `.strip()`.

Just for fun: https://unix.stackexchange.com/questions/276751/is-space-allowed-between-and-bin-bash-in-shebang

Reply:
Great catch, thanks.

I updated this to use `rstrip(b"\n")` to remove only the trailing newline